### ISSUE TITLE
Utils.encodeURI and RFC3986 ... 

### ISSUE THREAD

**Issue by pizzi80** [omnifaces/omnifaces#745]
The function `Utils.encodeURI` is it supposed to encode the input according to RFC 3986?

If it is not, would be useful to add that function in Utils, something like:

```
public static String encodeRFC3986(String string) {
  return string == null ? null : encodeURL(string).replace("+", "%20").replace("*", "%2A").replace("%7E", "~");
}

```
[https://stackoverflow.com/questions/5864954/java-and-rfc-3986-uri-encoding](url)

The php manual says that the php function `rawurlencode` encode the input according to RFC 3986.

I made some tests and the output it's equal to the function I'm proposing...  

[https://www.php.net/manual/en/function.rawurlencode.php](https://www.php.net/manual/en/function.rawurlencode.php)

```
#!/bin/php
<?php
$s = "Foo 'Bar (baz?) \"hello again\"'";

$encoded = urlencode($s); 
echo 'Encoded:         ',$encoded,PHP_EOL;

$encodedRFC3986 = rawurlencode($s); 
echo 'Encoded RFC3986: ',$encodedRFC3986,PHP_EOL;
?>
```

**Comment by BalusC** [comment]
Yes it's RFC3986.

I'll probably have to update javadoc and/or unit tests.

**Comment by BalusC** [comment]
On second inspection, it's actually RFC2396.

And now I'm not sure whether to update the existing method or to add a new one because 3986 obsoletes 2396.

**Comment by pizzi80** [comment]
I saw the 3986 version in many api code's examples during the last years, 
eg. Facebook rest api client, Amazon Product rest api client ... 

Due to this I have only the 3986 version in my "Utils" ... 

also I use Apache Commons Lang `StringUtils.replaceEach`
to optimize the replace without multiple intermediate String replace/allocation

[https://commons.apache.org/proper/commons-lang/apidocs/index.html](url)

don't know if on java 17+ it's still relevant, they finally improved the native `replace` a lot... :checkered_flag:

**Comment by BalusC** [comment]
Implemented for 4.2. I've just changed the encodeURI method.
